src/core/components/pin-set.js

```diff
@@ -0,0 +1,283 @@
+      })
+      pinSet.storeItems(items, logInternalKey, (err, rootNode) => {
+        if (err) { return callback(err) }
+        const opts = { cid: new CID(rootNode.multihash) }
+        dag.put(rootNode, opts, (err, cid) => {
+          if (err) { return callback(err) }
+          logInternalKey(rootNode.multihash)
+          callback(null, rootNode)
+        })
+      })
+    },
+
+    storeItems: (items, logInternalKey, callback, _depth, _subcalls, _done) => {
+      // callback (err, rootNode)
+      callback = once(callback)
+      // const seed = crypto.randomBytes(4).readUInt32LE(0, true)  // old nondeterministic behavior
+      const seed = _depth // new deterministic behavior
+      const pbHeader = pb.Set.encode({
+        version: 1,
+        fanout: defaultFanout,
+        seed: seed
+      })
+      let rootData = Buffer.concat([
+        Buffer.from(varint.encode(pbHeader.length)), pbHeader
+      ])
+      let rootLinks = []
+      for (let i = 0; i < defaultFanout; i++) {
+        rootLinks.push(new DAGLink('', 1, emptyKey))
+      }
+      logInternalKey(emptyKey)
+
+      if (items.length <= maxItems) {
+        // the items will fit in a single root node
+        const itemLinks = []
+        const itemData = []
+        const indices = []
+        for (let i = 0; i < items.length; i++) {
+          itemLinks.push(new DAGLink('', 1, items[i].key))
+          itemData.push(items[i].data || Buffer.alloc(0))
+          indices.push(i)
+        }
+        indices.sort((a, b) => {
+          const x = Buffer.compare(itemLinks[a].multihash, itemLinks[b].multihash)
+          if (x) { return x }
+          return (a < b ? -1 : 1)
+        })
+        const sortedLinks = indices.map((i) => { return itemLinks[i] })
+        const sortedData = indices.map((i) => { return itemData[i] })
+        rootLinks = rootLinks.concat(sortedLinks)
+        rootData = Buffer.concat([rootData].concat(sortedData))
+        DAGNode.create(rootData, rootLinks, (err, rootNode) => {
+          if (err) { return callback(err) }
+          return callback(null, rootNode)
+        })
+      } else {
+        // need to split up the items into multiple root nodes
+        // (using go-ipfs "wasteful but simple" approach for consistency)
+        _subcalls = _subcalls || 0
+        _done = _done || 0
+        const hashed = {}
+        const hashFn = (seed, key) => {
+          const buf = Buffer.alloc(4)
+          buf.writeUInt32LE(seed, 0)
+          const data = Buffer.concat([
+            buf, Buffer.from(toB58String(key))
+          ])
+          return fnv1a(data.toString('binary'))
+        }
+        // items will be distributed among `defaultFanout` bins
+        for (let i = 0; i < items.length; i++) {
+          let h = hashFn(seed, items[i].key) % defaultFanout
+          hashed[h] = hashed[h] || []
+          hashed[h].push(items[i])
+        }
+        const storeItemsCb = (err, child) => {
+          if (callback.called) { return }
```

Comment:
You can use http://npmjs.com/once for this

Reply:
Oops, looks like I had onced the callbacks but missed removing some of these.